### PR TITLE
Feature/dataset loader

### PR DESCRIPTION
**Pull request by ManiacMaxo**

### CODE REVIEW

File: nujo/utils/data/data_loader.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from os import mkdir
+from os.path import exists, expanduser
+import requests
+
+
+class DataLoader:
+    def __init__(self, name):
+        self._name = name
+        self._link = 'https://archive.ics.uci.edu/ml/machine-learning-databases/{0}/{0}.data'.format(
+            name)
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: - [x] @ManiacMaxo Use f-strings, not `.format` (f strings are faster and cooler 😃).  Check out how f-strings work [here](https://www.geeksforgeeks.org/formatted-string-literals-f-strings-python/).

- [x] Move `https://archive.ics.uci.edu/ml/machine-learning-databases/` in a separate file as a constant

File: nujo/utils/data/data_loader.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    def download(self):
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py:21]: @ManiacMaxo Assign `~/.nujo` to a separate variable as a constant and use it instead.

File: nujo/utils/data/dataset.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+from numpy import array, empty, float32, vstack
+from nujo.utils.data.dataset_iterator import DatasetIterator
+from os.path import expanduser
```
Comment (VIVelev) [inline nujo/utils/data/dataset.py]: @ManiacMaxo Those imports are not sorted 😁

File: nujo/utils/data/dataset_iterator.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+class DatasetIterator:
+    ''' Dataset Iterator
+
+    Parameters:
+    -----------
+    dataset : ....
+
+    '''
```
Comment (VIVelev) [inline nujo/utils/data/dataset_iterator.py]: @ManiacMaxo This docstring is garbage, please edit 😠

Comment (VIVelev) [reply nujo/utils/data/dataset_iterator.py]: Also, add some tiny docstring to the other classes

File: pyproject.toml [hunk 5]
```diff
@@ -15,6 +15,7 @@ pytest = "^5.3.4"
 flake8 = "^3.7.9"
 yapf = "^0.29.0"
 jupyter = "^1.0.0"
+requests = "^2.23.0"
```
Comment (VIVelev) [inline pyproject.toml]: @ManiacMaxo `requests` is not a dev dependency.
Use `poetry add requests`, not `poetry add requests --dev`

File: nujo/utils/data/data_loader.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        r = requests.get(self._link)
+        file = expanduser('~/.nujo/') + self._name + '.data'
+        nujo = expanduser('~/.nujo')
+        if not exists(nujo):
+            mkdir(nujo)
+            print("Directory '~/.nujo' Created ")
+        else:
+            print("Directory '~/.nujo' already exists")
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: Use only single quotes `'`, don't use double quote

File: nujo/utils/data/data_loader.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        print('File {} has been created.'.format(self.name))
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: Same, use f-strings

File: nujo/utils/data/data_loader.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        open(file, 'wb').write(r.content)
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: @ManiacMaxo use a `with` block when operating with files, no matter the size of the block.

File: nujo/utils/data/dataset.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        with open(file, 'r+') as data:
+            lines = data.readlines()
+            self.X = empty((0, 4))
+            self.y = empty((0, 1))
+
+            for line in lines[:-1]:
+                x = array([line.split(',')[:4]], dtype=float32)
+                self.X = vstack((self.X, x))
+                y = array([line.split(',')[-1][:-1]], dtype='U')
+                self.y = vstack((self.y, y))
```
Comment (VIVelev) [inline nujo/utils/data/dataset.py]: @ManiacMaxo those all can be out of the `with` block. In the `with` block put only what's necessary.

**Review by VIVelev — CHANGES_REQUESTED** [review]
💌

File: nujo/utils/data/dataset.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        with open(file, 'r+') as data:
+            lines = data.readlines()
+            self.X = empty((0, 4))
+            self.y = empty((0, 1))
+
+        for line in lines[:-1]:
+            x = array([line.split(',')[:4]], dtype=float32)
+            self.X = vstack((self.X, x))
+            y = array([line.split(',')[-1][:-1]], dtype='U')
+            self.y = vstack((self.y, y))
```
Comment (VIVelev) [inline nujo/utils/data/dataset.py]: Move this logic in a separate method. `__init__` methods should only contain variable assignments. Also, save `name` as public attribute.

Make it look something like:
```
def __init__(self, name):
        self.name = name
        ...
        self.X, self.y = self._load_from_file(filename)

def _load_from_file(self, filename):
        ...
```

File: nujo/utils/data/dataset.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    name : will be downloaded from the UCI ML repo
+
+    Returns:
+    --------
+    tuple : stores the csv dataset,
+    - np array with floating point integers
+    - np array with labels
+    '''
+    def __init__(self, name):
+        file = expanduser('~/.nujo/') + name
```
Comment (VIVelev) [inline nujo/utils/data/dataset.py]: That's why you should export it in a separate file and import it from the file when needed

File: nujo/utils/data/dataset.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+            y = array([line.split(',')[-1][:-1]], dtype='U')
+            self.y = vstack((self.y, y))
+
+    def __iter__(self):
+        return DatasetIterator(self)
+
+
+data = Dataset('iris')
+for line in data:
+    print(line)
```
Comment (VIVelev) [inline nujo/utils/data/dataset.py]: Put it in `if __name__ == '__main__':`

File: nujo/utils/data/data_loader.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+    Parameters:
+    -----------
+    name : will be downloaded from the UCI ML repo
+    '''
+    _UCI_REPO_URL = 'https://archive.ics.uci.edu/ml/machine-learning-databases/{}/{}.data'
+    _HOME_DIR = expanduser('~/.nujo')
+
+    def __init__(self, name):
+        self._name = name
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: Leave it public: `self.name` not `self._name`

File: nujo/utils/data/data_loader.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        self._line = self._UCI_REPO_URL.format(self._name)
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: Don't you mean `_link` not `_line` 😃

File: nujo/utils/data/data_loader.py (hunk 13 above)
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: Leave it public: `self.name` not `self._name`

File: nujo/utils/data/data_loader.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
+from os import mkdir
+from os.path import exists, expanduser
+
+import requests
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: `from requests import get`

File: nujo/utils/data/data_loader.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+            mkdir(self._HOME_DIR)
+            print('Directory "~/.nujo" Created ')
+        else:
+            print('Directory "~/.nujo" already exists')
+        print(f'File {self._name} has been created.')
+        with open(file) as f:
+            f.write(r.content)
+
+
+DataLoader('iris').download()
```
Comment (VIVelev) [inline nujo/utils/data/data_loader.py]: Put it in `if __name__ == '__main__': `

**Review by VIVelev — CHANGES_REQUESTED** [review]

File: nujo/utils/data/dataset.py (hunk 11 above)
Comment (ManiacMaxo) [reply nujo/utils/data/dataset.py]: what do you mean

Comment (VIVelev) [reply nujo/utils/data/dataset.py]: `expanduser('~/.nujo/')` is used both here and in data_loader.py, so its better if you make a separate file where you assign `expanduser('~/.nujo/')` to a const and import it from there when needed

**Review by shristov-elsys — APPROVED** [review]

**Pull request merged**